Enhance to report more accurate end to end operational latency for volume attach/detach

yuxiangqian opened the pull request:
**What type of PR is this?**
/kind failing-test
/kind feature

**What this PR does / why we need it**:
What:
1. Move volume operation end to end latency/error count metrics to util/metrics.go
2. Define operation starttime cache in util
3. Update pv-controller to use the new type
4. Enhance attach-detach-controller to record more accurate end to end latency

Why:
To have more accurate end to end operational latency for volume attach/detach. See #79397 for more details

**Which issue(s) this PR fixes**:
Fixes #79397

**Special notes for your reviewer**:
e2e testings are still in-process

**Does this PR introduce a user-facing change?**:
```release-note

Existing metrics: "storage_operation_duration_seconds" and "storage_operation_errors_total" are NOT impacted. Users relying on the existing metrics will see no impact.

Extend metrics: "volume_operation_total_seconds" and "volume_operation_total_errors" to cover volume attach/detach operations for end to end operation latency and total error counts respectively. Existing metrics described previously faces in-accuracy when there's transient error happened during attach/detach operation.

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yuxiangqian committed:
Move volume operation end to end latency/error count metrics to util/metrics.go
Define operation starttime cache in util
Update pv-controller to use the new type
Enhance attach-detach-controller to record more accurate end to end latency

add e2e tests to verify metrics, fix potential face condition in testing

enrich desired state of world to clean up operation timestamp cache when an operation is aborted

fix potential races caused by deleted pointer in map

yuxiangqian commented:
/sig storage

yuxiangqian commented:
@msau42 @saad-ali @gnufied

yuxiangqian commented:
/assign @msau42 @gnufied @saad-ali @jingxu97

yuxiangqian commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -137,6 +138,7 @@ type reconciler struct {
 	volumePluginMgr               *volumepkg.VolumePluginMgr
 	kubeletPodsDir                string
 	timeOfLastSync                time.Time
+	operationStartTimeCache       util.OperationStartTimeCache
```
I will need to add comments here to explain why kubelet needs a bit change.
Confirmed with @msau42 kubelet attaching/detaching workflow should be deprecated.
This newly introduced cache here should be removed when the mentioned work has been done

yuxiangqian replied:
added e2e test

msau42 commented:
/remove-kind failing-test

oleksiys commented on pkg/volume/testing/testing.go:
```diff
@@ -1530,3 +1533,132 @@ func (f *fakeVolumeHost) SetKubeletError(err error) {
+	if loaded {
+		oldTs, ok := obj.(fakeOperationTimestamp)
+		if ok {
+			oldTs.countAddIfNotExist += 1
```
There's a race here. It might be easier to just use a regular map with a mutex, it's easier to reason about and we don't need throughput optimizations for testing.

yuxiangqian replied:
used a regular map + rwlock as suggested. thnx

oleksiys commented on pkg/volume/testing/testing.go:
```diff
@@ -1530,3 +1533,132 @@ func (f *fakeVolumeHost) SetKubeletError(err error) {
 	return nil
 }
+
+type FakeOperationStartTimeCache struct {
```
An alternative to make this public might be adding Keys() method on the actual interface.

yuxiangqian commented:
e2e tests metrics dumping gives the expected results as following, still working on enhance e2e tests to include the added cases.
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="0.1"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="0.25"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="0.5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="1"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="2.5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="10"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="15"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="25"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="50"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="120"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="300"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="600"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450",le="+Inf"} 1
volume_operation_total_seconds_sum{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450"} 0.033759335
volume_operation_total_seconds_count{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-hostpath-provisioning-450"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="0.1"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="0.25"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="0.5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="1"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="2.5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="5"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="10"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="15"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="25"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="50"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="120"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="300"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="600"} 1
volume_operation_total_seconds_bucket{operation_name="volume_attach",plugin_name="kubernetes.io/csi:csi-mock-csi-mock-volumes-2996",le="+Inf"} 1

yuxiangqian commented:
/test pull-kubernetes-integration

yuxiangqian commented:
integration testing kept failing on timeout. see lots of similar failures on Prow. [Failures](https://prow.k8s.io/?job=pull-kubernetes-integration&state=failure)

msau42 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -326,6 +328,25 @@ type attachDetachController struct {
+	//     start time: the first time "AttachVolume" is called in reconciler in "attachDesiredVolumes"
+	//                 for the VolumeToAttach object
+	//     end time:   after the "VolumeToAttach" has been successfully attached to the desired node
+	//     abort:      The "VolumeToAttach" has been marked "attached", i.e., actualStateOfWorld
```
We should also abort when the volume is removed from desired state.  This can happen if the pod is deleted before we finished attaching.

yuxiangqian replied:
good point. another possible scenario is when a pod has been updated to remove a volume. Will enhance to handle those scenarios.

msau42 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -326,6 +328,25 @@ type attachDetachController struct {
+	//                 for the "AttachedVolume" object
+	//     end time:   after the volume "AttachedVolume" object refers to being successfully
+	//                 detached from the node
+	//     abort:      N.A.
```
@gnufied @jingxu97 is it possible to "abort" detach if the volume gets added to desired state again?

yuxiangqian replied:
I think it is possible. the detach process first get the actual state, and cross compare to desired state, it will only do the detach operation if the volume does not appear in desired state. I will add logic to abort the operation under that circumstance to avoid possible memory leak. Any other scenarios? @gnufied @jingxu97

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -279,18 +293,26 @@ func (rc *reconciler) attachDesiredVolumes() {
 			}
 		}
 
+		// insert a detach operation start time into cache if needed for the volume for metrics reporting
```
typo: detach should be attach?

yuxiangqian replied:
noted

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -137,6 +138,13 @@ type reconciler struct {
+	// still exists in kubelet. The following cache is needed to make sure the new
+	// changes do not break build, though needs to be cleaned up when attach/detach
+	// are completely removed from kubelet
+	operationStartTimeCache util.OperationStartTimeCache
```
Or can it handle nil cache?

yuxiangqian replied:
adding checks in "VolumeAttach" call will make it capable of handling nil. Though IMHO, I'd rather make it explicit to avoid those unnecessary checks in the feature. Its an empty map which should not bring any burden to kubelet. I can add nil checking if its a more standard way of handling things in K8s though. WDYT?

msau42 replied:
IIUC, for kubelet attach/detach, we don't set any start time, so the record is a no-op?

yuxiangqian replied:
that is right. Since kubelet workflow is a to-be-deprecated workflow, I think it makes sense to be a no-op? let me know if you think the otherwise

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -236,6 +246,8 @@ func (rc *reconciler) reconcile() {
 				// Log all other errors.
 				klog.Errorf(attachedVolume.GenerateErrorDetailed("attacherDetacher.DetachVolume failed to start", err).Error())
+				// report error count metrics
+				util.RecordMetric(createOperationUniqueKey(attachedVolume.VolumeName, attachedVolume.NodeName), rc.operationStartTimeCache, err)
```
We already record an error metric when we call OperationCompleteHook(). Are we going to be double counting errors with this?

yuxiangqian replied:
no we are not, it will go to another error metric: volume_operation_total_errors which is paired with volume_operation_total_seconds for the sake of consistency

msau42 commented:
I did an initial quick review. Will continue later.

gnufied commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -279,18 +293,26 @@ func (rc *reconciler) attachDesiredVolumes() {
+		// note that the plugin name has been set to "" as there is no visibility at this moment to which
+		// plugin/csi driver would be conducting the operation. It will be populated later when detach
+		// operation function is created and plugin name then could be decided
+		rc.operationStartTimeCache.AddIfNotExist(createOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName), "", "volume_attach")
```
This does not cover end to end operational metrics as it should. I would expect, end to end metrics to cover time from which attach was request to time when attach was finished. This metric does not include time for:

1. Volume to detach from previous node if an operation was pending
2. Time waiting for reconciler to kick the operation in.

As such - I think the numbers reported here will be very close to already pre-existing operation metrics and hence aren't very useful as brand new metrics.

yuxiangqian replied:
for
1. my understanding is there are two cases:
    a. there is an attach operation pending to be finished, which is fine not to add anything into the cache
    b. there is a detach operation needs to happen before the attach can be kicked off. There should be a detach metric being recorded in this case, and getting ride of that piece from the whole attach operation process seems to be also reasonable to me, the end of day the metric reports end to end latency of attach operation.
  2. "Time waiting for reconciler to kick the operation in" I assume you meant the time elapsed until "attachDesiredVolumes" will be called? Now the question will still be will that new metric should record controller related delay or not. If not, ignoring the time elapsed before an actual attach operation is valid.

for the last piece, It will only be close to already existed operation metrics if there is NO error happened during the whole process. If there's any error, the real start time of attach operation will be gone along with goroutine finishes "GenerateAttachVolumeFunc" returned "onComplete" call, which is a major fix of this PR

yuxiangqian replied:
discussed offline. The metric will not include the waiting time when there's a pending operation on a specific volume for the sake of eliminating confusions and to simply SLA/alert set up for end users

yuxiangqian commented:
based on conversation with @msau42 @gnufied, enriched desired state of world to clean up cache when a volume attach/detach operation should be aborted.

yuxiangqian commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

msau42 commented on pkg/volume/testing/testing.go:
```diff
@@ -1530,3 +1533,123 @@ func (f *fakeVolumeHost) SetKubeletError(err error) {
+	rwLock  sync.RWMutex
+}
+
+func newFackOperationTimestamp(plugin, op string) *fakeOperationTimestamp {
```
fake :-)

yuxiangqian replied:
woops, fixed.

msau42 commented on pkg/volume/util/metrics.go:
```diff
@@ -63,6 +63,13 @@ var storageOperationEndToEndLatencyMetric = prometheus.NewHistogramVec(
+var volumeOperationErrorsMetric = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "volume_operation_total_errors",
+		Help: "Total volume operation erros",
```
Was this a previous metric that existed before and was being consumed?

The general feedback I've gotten from sig-instrumentation is that total errors is not useful without a total count.  So I'm not sure if we want to continue to support this metric. We currently have 3 different metrics that report errors, and I would like to deprecate 2 of them and only continue to support 1 going forward.

yuxiangqian replied:
No it does not. The metric, "volume_operation_total_errors" and "volume_operation_total_seconds" only exists for "provision" and "deletion" operations. For "attach" and "detach", it's newly introduced.
It's a very legit feedback. The idea is once "volume_operation_total_seconds" have been fully implemented to cover all volume operations, mount/unmount, resizing, snapshot etc, it should replace the current "storage_operation_seconds".
For the error one, it's a bit tricky, there are two things to keep in mind, one is naming, if we have decided to use "volume_operation_total_seconds", it makes sense to have the error metric named as volume_operation_xxx_errors than re-using current ones. The other would be the meaning of total errors. The retry logic in controllers makes total error count a bit different to a typical service failure call. It's totally possible the total error count being larger than total count of operations (assuming total count means total number of operations). In this case, having total error counts only seems to be not a horrible idea.

msau42 replied:
Maybe we should take a step back and try to come up with a clear long-term model of how we want the volume metrics to look like.  For me, the picture is very confusing right now because:
* we have K8s style reconciler loops which don't quite match up with traditional service operations. 
* csi and in-tree are different
* we're trying to derive metrics for all these on the client-side, which maybe doesn't make sense when considering csi

msau42 replied:
Conclusion from offline discussion is let's just focus on e2e latency for now.  We will followup with a future proposal/KEP on how to count errors at various levels: plugin, controller, user.

msau42 replied:
Also, there's a typo in the help msg, should be "errors"

yuxiangqian replied:
fixed. thanks

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -444,11 +449,15 @@ func (og *operationGenerator) GenerateAttachVolumeFunc(
 
+	// update operation start time cache with fully qualified plugin name for the corresponding volumeToAttach
+	fullPluginName := util.GetFullQualifiedPluginNameForVolume(attachableVolumePluginName, volumeToAttach.VolumeSpec)
+	operationStartTimeCache.UpdatePluginName(createOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName), fullPluginName)
```
/assign @davidz627 
to check that the plugin name is correct here in the case of migration.

davidz627 replied:
this is correct for migration, since it takes the `VolumeSpec` and the `attachableVolumePluginName` from the above block where we get the correct values for those

yuxiangqian replied:
thanks @davidz627

yuxiangqian commented on pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:
```diff
@@ -195,6 +210,15 @@ type pod struct {
+// cleanCacheOnOperationAbort deletes cached operation timestamp of a specific
+// volume-node pair from cache in the case when the operation should be aborted
+func cleanCacheOnOperationAbort(volumeName v1.UniqueVolumeName, nodeName k8stypes.NodeName, toBeCleanedOperationName string, operationTimestamps util.OperationStartTimeCache) {
+	key := string(nodeName) + "/" + string(volumeName)
```
I have a function defined in reconcile.go, named as "createOperationUniqueKey". I guess I can move it to the local util.go file. I will do that.

yuxiangqian commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -258,6 +270,8 @@ func (rc *reconciler) attachDesiredVolumes() {
 			}
 			rc.actualStateOfWorld.ResetDetachRequestTime(volumeToAttach.VolumeName, volumeToAttach.NodeName)
+			// remove any pre-existed cache entry in this case. The volume has been attached
+			rc.operationStartTimeCache.Delete(createOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName))
```
It should not however there is a bug in the implementation I just fixed based on this.
The actualStateOfWorld will only be updated after the metric has been reported in operation_generator.go
I am adding the deletion here just to clean up the cache to avoid potential dead entry when the following situation happens:
1. controller error'd out trying to attach the volume to the node
2. volume has been attached to the node out of the bound
3. controller found the volume has been attached and stopped calling Attach
IMHO, the metric should not be reported as it may introduce unnecessary noises, however the cache should have been cleaned.

yuxiangqian commented:
@msau42 changes have been made based on your feedback, but I am holding up on another push before I collected all feedbacks.

davidz627 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -219,11 +223,17 @@ func (rc *reconciler) reconcile() {
+			// note that the plugin name has been set to "" as there is no visibility at this moment to which
+			// plugin/csi driver would be conducting the operation. It will be populated later when detach
+			// operation function is created and plugin name then could be decided
+			rc.operationStartTimeCache.AddIfNotExist(createOperationUniqueKey(attachedVolume.VolumeName, attachedVolume.NodeName), "", "volume_detach")
```
It's hard to understand/maintain when the start of the metric exists here (in the reconciler) and I believe the "end" of the same metric is after the success in the operation generator. Could we also cache the start time in the start of the function in the operation generator and just have it not overwrite the start time if it is on the n'th run-through? 

I'm also not exactly sure what the goal is so maybe i'm misunderstanding :)

/cc @msau42

yuxiangqian replied:
The goal is to record end to end attach/detach operation. However, if a detach operation needs to happen before an attach can happen, i.e., a volume is currently attached to a different node when the controller tries to attach it to a specific node, the detach operation latency will NOT be included the attach operation latency.
You are not misunderstanding, and the end of the operation would be when the final success happens in operation generator. It is also possible to have the cache set the first time an attach/detach operation has been generated in operation generator.
Putting it in reconcile though, makes it easier to manage the lifecycle of the cached item, i.e., when an attach/detach operation needs to be aborted while trying. Also this is also more extendable in the future if we'd like to report "waiting_time" metrics, i.e., the total time has spend on waiting on a volume to be detached before attach can happen.

The cached start timestamp will eventually be removed when Operation generator's attach function returns success, and thats when the metric will be reported.

yuxiangqian replied:
@davidz627 I have moved the insertion into operation generator as the time_waiting piece and error reporting still remains to be discussed. Please take a look. thnx

yuxiangqian commented:
Had an offline conversation with @msau42 @saad-ali. For the PR, I am going to focus on implementation of having end to end latency metrics for attach-detach operation.
Error metric reporting will be resolved in future enhancement, and I will remove any error metric recording in this PR.
Discussions also have been conducted on whether the end to end latency of attach operation should include potential detach operation needs to happen before an attach can happen. The conclusion is to not include to avoid confusion, however a separate one which records the "waiting_to_attach" could be useful (out of the scope for this PR).

yuxiangqian committed:
remove error metric reporting, define common function to create cache key, fix typo
remove unused function
moving cache timestamp insertion into operation generator for attach/detach

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -219,11 +223,17 @@ func (rc *reconciler) reconcile() {
+			// note that the plugin name has been set to "" as there is no visibility at this moment to which
+			// plugin/csi driver would be conducting the operation. It will be populated later when detach
+			// operation function is created and plugin name then could be decided
+			rc.operationStartTimeCache.AddIfNotExist(operationexecutor.CreateAttachDetachOperationUniqueKey(attachedVolume.VolumeName, attachedVolume.NodeName), "", "volume_detach")
```
do we need to start the time now, or can we wait until inside operation generator when we actually have the plugin name?

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -258,6 +268,8 @@ func (rc *reconciler) attachDesiredVolumes() {
 			}
 			rc.actualStateOfWorld.ResetDetachRequestTime(volumeToAttach.VolumeName, volumeToAttach.NodeName)
+			// remove any pre-existed cache entry in this case. The volume has been attached
+			rc.operationStartTimeCache.Delete(operationexecutor.CreateAttachDetachOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName))
```
should it be abort, or should we actually log the time?  What if attach operation starts, times out, and next time before we retry, it finishes?

yuxiangqian replied:
It's abort here to avoid leaking in cases when the volume has been attached out of bound while the controller is trying.
for the case "What if attach operation starts, times out, and next time before we retry, it finishes?", it will be a no-op. The operation-generator change reports the metrics before asw is updated.

msau42 replied:
Are we sure that ASW cannot be updated through some other operation like VerifyVolumesAreAttached?

Would it be safer to just record metric anyway?  If it was already recorded by operation generator, then this should be a noop right?

msau42 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -279,11 +291,21 @@ func (rc *reconciler) attachDesiredVolumes() {
+		//       Including the "detach" time might introduce challenges to define SLO or to set alerts based on it.
+		//       Whether the "detach" operation will happen depends on the current status of the volume (currently attached to any none-desired node or not)
+		//       and "detach" can be a long operation.
+		rc.operationStartTimeCache.AddIfNotExist(operationexecutor.CreateAttachDetachOperationUniqueKey(volumeToAttach.VolumeName, volumeToAttach.NodeName), "", "volume_attach")
```
same thing here, can we start the timer once we know the plugin? The logic will be simpler to follow if everything is initialized at the same time.

yuxiangqian replied:
we could certainly do that as @davidz627 mentioned previously [here](https://github.com/kubernetes/kubernetes/pull/80832#discussion_r314907693)
I prefer to put in to reconciler instead just for the sake of covering "waiting_time" in the future, such that the cache management happens within the same unit.
The other reason, which does not hold for now but might hold later, is for error metric reporting. Operation generator returns error in many cases (missing plugin could be one of it), when should it add the start timestamp into cache? If put after a valid plugin is found, the cache will not have the entry and makes error metric reporting tricky.
thoughts?

msau42 replied:
Right now cache management is not in the same component. We're starting it in reconciler, and finishing it in operation generator.

I think we'll need to revisit error reporting/categorization.  Plugin-level errors can be tracked per operation, and we can get error percentage out of that.  I don't think higher level errors at the controller level should be tracked with the same metric, as there's not a great notion of error percentage.

yuxiangqian replied:
when I say management, I actually meant adding into cache for "waiting_time". the deletion along with metric reporting needs to happen in operation generator because of various of reasons.
yeah, error reporting/categorization needs to be planned out before it is added for sure.

msau42 replied:
Hm that means for the metric that measures time waiting for detach before attaching, the plugin name will be empty? Is that odd? @gnufied 

I would rather keep this logic as simple as possible and not have a possibility of plugin name to be empty, and have to make 2 calls in 2 different places to actually initialize the metric.

gnufied replied:
It will be odd indeed to emit the metric without the plugin name for "waiting_for_detach" case. The only reason we do not have plugin name here is because we have not actually inspected the volume right? This can be fixed later I think.

yuxiangqian replied:
@gnufied I have updated the PR according to the feedback, anyways I am not including "waiting_for_detach" within this PR.

yuxiangqian commented:
for error metric, it exists already in [persistentvolume/metrics.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/metrics/metrics.go#L95). It is moved over the a lower level package along with the change and not used by attach/detach operation. However it is currently used by provision/deletion which is not ideal. Moving it together with volume_operation_total_seconds makes sense to me for later on clean up if needed.

gnufied commented on pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:
```diff
@@ -130,10 +130,11 @@ type PodToAdd struct {
-		volumePluginMgr: volumePluginMgr,
+		nodesManaged:        make(map[k8stypes.NodeName]nodeManaged),
+		volumePluginMgr:     volumePluginMgr,
+		operationTimestamps: operationTimestamps,
```
the `operationTimestamps` variable is effectively a package global - will it be cleaner to initialize it as such and have all the code access it directly?

yuxiangqian replied:
Certainly that is an option I have considered. The major thing I was not doing that was the complexity of guarantee key uniqueness across the board with all operations. i.e., provisioning/deletion v.s. attach/detach. It would be a great option if key generation could be managed in a central place as well, however I could not find an easy way or place to put it. WDYT?

yuxiangqian commented:
/test pull-kubernetes-conformance-kind-ipv6

yuxiangqian commented:
/test pull-kubernetes-conformance-kind-ipv6

yuxiangqian commented:
ipv6 testing seems to be flaky at this point: [test-grid](https://testgrid.k8s.io/sig-testing-kind#pull-kubernetes-conformance-kind-ipv6)
